Reporting/ad-extensions-reporter.js

```diff
@@ -0,0 +1,395 @@
+    7: 'PROMOTION START',
+    8: 'PROMOTION END',
+    9: 'OCCASION',
+    10: 'FINAL URLS',
+    11: 'FINAL MOBILE URLS',
+    12: 'TRACKING URL',
+    13: 'LANGUAGE',
+    14: 'FINAL URL SUFFIX'
+}
+
+var extensionPlaceholderTypeIdToFieldIdsObjectMapping = {
+    1: sitelink,
+    2: call,
+    3: app,
+    7: location,
+    17: callout,
+    24: structuredSnippet,
+    30: affiliateLocation,
+    31: message,
+    35: price,
+    38: promotion
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+// Functions
+
+function main() {
+    var spreadsheet = checkSpreadsheet(spreadsheetUrl, "Output Spreadsheet");
+    var sheets = spreadsheet.getSheets();
+
+    Logger.log("Preparing spreadsheet");
+    try {
+        allExtensionsSheet = spreadsheet.getSheetByName("All Extensions").clear();
+    } catch (e) {
+        allExtensionsSheet = spreadsheet.insertSheet("All Extensions");
+    }
+
+    var extensionTabNames = Object.keys(extensionPlaceholderTypeIdToFieldIdsObjectMapping).map(
+        function (key) { return extensionPlaceholderTypeIdToFieldIdsObjectMapping[key].name }
+    )
+    for (var i = 0, n = sheets.length; i < n; i++) {
+        if (extensionTabNames.indexOf(sheets[i].getName()) !== -1) {
+            spreadsheet.deleteSheet(sheets[i]);
+        }
+    }
+
+    // Get report
+    Logger.log("Getting report");
+    if (typeof includeExtensionsWithZeroImpressions === 'undefined') {
+        Logger.log("Warning: options variable includeExtensionsWithZeroImpressions not found. " +
+            "Continuing, and by default including extensions that have had no impressions today."
+        );
+        impressionsCondition = '';
+    } else {
+        if (includeExtensionsWithZeroImpressions === true) {
+            var impressionsCondition = '';
+        } else if (includeExtensionsWithZeroImpressions === false) {
+            var impressionsCondition = 'WHERE Impressions > 0 ';
+        } else {
+            throw ("Value of options variable includeExtensionsWithZeroImpressions must be a boolean, either true or false. " +
+                "Currently, it is: " + includeExtensionsWithZeroImpressions)
+        }
+    }
+
+    var placeholderFeedItemReport = AdsApp.report(
+        'SELECT PlaceholderType, AttributeValues, DisapprovalShortNames ' +
+        'FROM PLACEHOLDER_FEED_ITEM_REPORT ' +
+        impressionsCondition +
+        'DURING TODAY'
+    );
+    var reportRows = placeholderFeedItemReport.rows();
+
+    // Process report data, to prepare for outputting to the spreadsheet
+    Logger.log("Processing report data");
+    var extensionPlaceholderTypeIds = Object.keys(extensionPlaceholderTypeIdToFieldIdsObjectMapping);
+
+    var extensionTypeFeedItemCounts = {};
+    for (var i = 0, n = extensionPlaceholderTypeIds.length; i < n; i++) {
+        extensionTypeFeedItemCounts[extensionPlaceholderTypeIds[i]] = 0;
+    }
+    var jsonAttributeValuesByExtensionType = {};
```

Comment:
Your variable names are very descriptive, which is helpful because I know what is in them by reading the name, but also makes the code a bit harder to read sometimes, as they make the lines really long, and makes it hard for me to consistently recognise the variable name (as it is hard to remember the whole thing!)

If you are using a variable a lot, sometimes it is better to have a shorter name that is easier for a human to remember, and recognise throughout the code. e.g. `extensions` or `extensionsByType`?

Reply:
Yes that makes a lot of sense! It's also hard for me to remember what each variable is, and I think that slowed me down when writing the script. I've shortened some variable names.